internal/controller/resource_manager.go

```diff
@@ -144,74 +147,124 @@ func getErrFromWatchEvent(event watch.Event) error {
 // watchEvents watches Kubernetes resources. The second part of the of the "list
 // and watch" strategy.
 // Any occurring errors are sent through errChanWatch. The watch stops when ctx is cancelled.
-func (r *ResourceManager) watchEvents(ctx context.Context, apiGroup string, watchOptions metav1.ListOptions, mutationChan chan *pb.KubernetesResourceMutation) error {
-	logger := r.logger.With(zap.String("api_group", apiGroup))
+func (r *ResourceManager) watchEvents(ctx context.Context, resourceVersion string, mutationChan chan *pb.KubernetesResourceMutation) error {
+	logger := r.logger
 
-	objGVR := schema.GroupVersionResource{Group: apiGroup, Version: "v1", Resource: r.resourceName}
+	// Build watch options from the provided resource version
+	watchOptions := metav1.ListOptions{
+		Watch:               true,
+		ResourceVersion:     resourceVersion,
+		AllowWatchBookmarks: true,
+	}
 
-	watcher, err := r.dynamicClient.Resource(objGVR).Namespace(metav1.NamespaceAll).Watch(ctx, watchOptions)
-	if err != nil {
-		logger.Error("Error setting up watch on resource", zap.Error(err))
+	lastKnownResourceVersion := resourceVersion
 
-		return err
-	}
+	var watcher watch.Interface
+
+	defer func() {
+		if watcher != nil {
+			watcher.Stop()
+		}
+	}()
 
 	mutationCount := 0
 
+	for {
+		// (Re)initialize the watcher using a helper
+		w, err := r.startWatcher(ctx, r.apiGroup, lastKnownResourceVersion, watchOptions, watcher, logger)
+		if err != nil {
+			return err
+		}
+
+		watcher = w
+
+		// Run the watcher loop until a restart is requested or an error occurs
+		updatedRV, restart, err := r.runWatcherUntilRestart(ctx, watcher, lastKnownResourceVersion, mutationChan, logger, &mutationCount)
+		if err != nil {
+			return err
+		}
+
+		lastKnownResourceVersion = updatedRV
+
+		if !restart {
+			return nil
+		}
+	}
```

Comment:
Merge `startWatcher` and `runWatcherUntilRestart`. I don't see any reason to artificially decouple them?

```suggestion
	for {
		// Run the watcher loop until a restart is requested or an error occurs
		lastKnownResourceVersion, restart, err := r.runWatcherUntilRestart(ctx, lastKnownResourceVersion, mutationChan, logger, &mutationCount)
		if err != nil {
			return err
		}

		if !restart {
			return nil
		}
	}
```